// udn-frontend/index.ts
export interface Message {
  // connecting to server
  requestingServerConnection?: boolean;

  // mailbox
  requestingMailboxSetup?: boolean;
  assignedMailboxId?: string; // sent by server
  requestedMailbox?: string;

  // subscribing to channel
  subscribeChannel?: string;
  unsubscribeChannel?: string;
  subscribed?: boolean; // sent by server to confirm subscription

  //sending message
  messageChannel?: string;
  messageBody?: string;
}

export default class UDNFrontend {
  ws: WebSocket | undefined;

  // handlers
  private connectionHandler = () => {};
  private disconnectionHandler = () => {};
  private messageHandler = (data: Message) => {};
  private mailboxHandler = (mailboxId: string) => {};

  // init
  set onconnect(handler: () => void) {
    this.connectionHandler = handler;
  }
  set ondisconnect(handler: () => void) {
    this.disconnectionHandler = handler;
  }
  set onmessage(handler: (data: Message) => void) {
    this.messageHandler = handler;
  }
  set onmailbox(handler: (mailboxId: string) => void) {
    this.mailboxHandler = handler;
  }

  // utility methods
  private send(messageObject: Message): boolean {
    if (this.ws == undefined) return false;

    const messageString = JSON.stringify(messageObject);
    this.ws.send(messageString);
    return true;
  }

  // public methods
  connect(address: string): void {
    this.disconnect();

    this.ws = new WebSocket(address);
    this.ws.addEventListener("open", this.connectionHandler);
    this.ws.addEventListener("close", this.disconnectionHandler);
    this.ws.addEventListener("message", (message) => {
      const dataString = message.data.toString();
      const data = JSON.parse(dataString);

      if (data.assignedMailboxId) {
        return this.mailboxHandler(data.assignedMailboxId);
      } else {
        this.messageHandler(data);
      }
    });
  }

  disconnect(): void {
    this.ws?.close();
  }

  sendMessage(channel: string, body: string): boolean {
    const messageObject = {
      messageChannel: channel,
      messageBody: body,
    };
    return this.send(messageObject);
  }

  subscribe(channel: string): boolean {
    const messageObject = {
      subscribeChannel: channel,
    };
    return this.send(messageObject);
  }

  unsubscribe(channel: string): boolean {
    const messageObject = {
      unsubscribeChannel: channel,
    };
    return this.send(messageObject);
  }

  requestMailbox(): boolean {
    const messageObject = {
      requestingMailboxSetup: true,
    };
    return this.send(messageObject);
  }

  accessMailbox(mailboxId: string): boolean {
    const messageObject = {
      requestedMailbox: mailboxId,
    };
    return this.send(messageObject);
  }
}
